Update benchmark test to actively used solvers (#69)
PTsolvers/JustRelax.jl: README.md

Need to solve a very large multi-physics problem on a GPU cluster? Just Relax!

Pseudo-transient accelerated iterative solvers, ready for extreme-scale, multi-GPU computation.

`JustRelax.jl` is a collection of pseudo-transient relaxation solvers
for multi-physics problems on regular, staggered, parallel grids,
using MPI and multiple CPU or GPU backends. It's part of the [PTSolvers project](https://ptsolvers.github.io) and
the [GPU4GEO project](https://www.pasc-ch.org/projects/2021-2024/gpu4geo/). Current publications, outreach and news can be found on the [GPU4GEO website](https://ptsolvers.github.io/GPU4GEO/).

The package relies on other packages as building blocks and parallelisation tools:

* [ParallelStencil.jl](https://github.com/omlins/ParallelStencil.jl)
* [ImplicitGlobalGrid.jl](https://github.com/omlins/ImplicitGlobalGrid.jl)
* [GeoParams.jl](https://github.com/JuliaGeodynamics/GeoParams.jl)
* [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl)

The package serves several purposes:

>   * It reduces code duplication between several applications, e.g. [PseudoTransientStokes.jl](https://github.com/PTsolvers/PseudoTransientStokes.jl). 

Reviewer: Maybe this point it not necessary in the README

  * It provides a collection of solvers to be used in quickly developing new applications
  * It provides some standardization so that application codes can

>      - more easily "add more physics" through the use of [GeoParams.jl]((https://github.com/JuliaGeodynamics/GeoParams.jl))

Reviewer: GeoParams doesn't really "add more physics", _just_ handles local material properties

     - more easily switch between a psuedo-transient solver and another solver (e.g. an explicit thermal solvers)

  * It provides a place to describe performance benchmarks for the solver routines
  * It provides a natural location for contributions of new solvers for use by the larger community

We include several miniapps, each designed to solve a well-specified benchmark problem, in order to provide

  - examples of high-performance usage,
  - bases on which to build more full-featured application codes
  - cases for reference and performance tests

> `JustRelax.jl` is used in the following applications:

Reviewer: If you want, you could link `MTK`. Buy maybe it's better to delete this for now, and add it when you have some more mature results?

  * :tools: TODO link to all applications using the package here (crucial for early development)

## Installation

`JustRelax.jl` is not yet registered (we are in the process), however it can be installed by cloning the repository. If you have navigated the terminal to the directory you cloned `JustRelax.jl`, you can test the package by running the following commands:

```julia
using JustRelax
julia> ] 
  pkg> activate .
  pkg> instantiate
  pkg> test JustRelax
```
The test will take a while, so grab a coffee or tea. 

## Usage

> The package relies on [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl) for the particle in cell method. Depending on where you want to run `JustRelax.jl` (CPU or GPU) you will need to specify the Backend. Run this command in the REPL and restart Julia:

Reviewer: ~~The package relies on JustPIC.jl for the particle in cell method~~ "For the advection of physical fields, this package relies on the Particles-In-Cell method implemented in JustPIC"

```julia
  set_backend("Threads_Float64_2D")       #running on the CPU
  set_backend("CUDA_Float64_2D")          #running on an NVIDIA GPU
  set_backend("AMDGPU_Float64_2D")        #running on an AMD GPU
```

> After you have run your desired line and restarted Julia, there should be a file called `LocalPreferences.toml` in the directory together with your `Project.toml` and `Manifest.toml`. This file contains the information about the backend you want to use. If you want to change the backend, you can run the command again. 

Reviewer: Add a comment at the end saying that CPU 2D is the default backend.

As stated above, the parallelisation is done using ParallelStencil.jl. Therefore, `JustRelax.jl` has as an environment already setup with which you can specify the dimension of your problem (2D or 3D) and the backend (CPU or GPU). The following commands are available:

```julia
  model = PS_Setup(:Threads, Float64, 2)  #running on the CPU in 2D
  environment!(model)

  model = PS_Setup(:CUDA, Float64, 2)     #running on an NVIDIA GPU in 2D
  environment!(model)

  model = PS_Setup(:AMDGPU, Float64, 2)   #running on an AMD GPU in 2D
  environment!(model)

> ```

Reviewer: Add comment here clarifying that `2` is the dimension, and `3` is for 3D

## Shear Band 2D example

This example benchmark test displays how the package can be used to simulate shear bands. The example is based on the [ShearBands2D.jl](miniapps/benchmarks/stokes2D/shear_band/ShearBand2D.jl). 

> The `CUDA.allowscalar(false)` is needed to avoid the use of scalar values in the code which would run slowly on a CPU and is disallowed on GPU systems such as HPC clusters. 

Reviewer: What `CUDA.allowscalar(false)`  does is to disallow scalar indexing. E.g. `A[1,1,1]` will throw an error if `A isa Union{CuArray, ROCArray}`.  This is indeed to make the code crash instead of accidentally having scalar indexing somewhere and running slower than a turtle (on CUDA; AMD will crash). I think this is a implementation/debugging detail which perhaps is not really relevant and we should just omit and actually remove it from the miniapps.

```julia
using CUDA
CUDA.allowscalar(false)

using GeoParams, GLMakie, CellArrays
using JustRelax, JustRelax.DataIO
```
Setup of environment and general solution for the benchmark test.

```julia

> # setup ParallelStencil.jl environment

Reviewer: remove this comment

model  = PS_Setup(:CUDA, Float64, 2)
environment!(model)

> # HELPER FUNCTIONS ---------------------------------------------------------------

Reviewer: remove these comments

solution(ε, t, G, η) = 2 * ε * η * (1 - exp(-G * t / η))

```

> The function `init_phases!` initializes the phases within cell arrays. The function is parallelized with `@parallel_indices (i,j) ` and `@parallel (JustRelax.@idx ni)`. In this case, this is the only function that needs to be taylored to the specific problem, everything else is handled by `JustRelax.jl` itself. 

Reviewer: "...is parallelized with the `@parallel_indices`  macro from ParallelStencil.jl."

```julia

> # Initialize phases on the particles

Reviewer: remove comment

function init_phases!(phase_ratios, xci, radius)
    ni      = size(phase_ratios.center)
    origin  = 0.5, 0.5
    
    @parallel_indices (i, j) function init_phases!(phases, xc, yc, o_x, o_y)
        x, y = xc[i], yc[j]
        if ((x-o_x)^2 + (y-o_y)^2) > radius^2
            JustRelax.@cell phases[1, i, j] = 1.0
            JustRelax.@cell phases[2, i, j] = 0.0
        
        else
            JustRelax.@cell phases[1, i, j] = 0.0
            JustRelax.@cell phases[2, i, j] = 1.0
        end
        return nothing
    end

>     @parallel (JustRelax.@idx ni) init_phases!(phase_ratios.center, xci..., origin...)

Reviewer: Could you add 'radius' as an argument of inner function?

The inner `init_phases!` is a closure, and it not may be obvious to everyone why `radius` is not being passed as an argument.

end
```

> Initialisation of the physical domain and the grid. As `JustRelax.jl` is utilising [ImplicitGlobalGrid.jl](https://github.com/omlins/ImplicitGlobalGrid.jl), the grid can be `MPIAWARE` through setting the grid steps in x- and y- direction to ` di = @. li / (nx_g(),ny_g())`. This makes it a global grid and the grid steps are automatically distributed over the MPI processes. 

Reviewer: ~~is utilising~~ employs / relies on

```julia

# MAIN SCRIPT --------------------------------------------------------------------
function main(igg; nx=64, ny=64, figdir="model_figs")

    # Physical domain ------------------------------------
    ly       = 1e0          # domain length in y
    lx       = ly           # domain length in x
    ni       = nx, ny       # number of cells
    li       = lx, ly       # domain length in x- and y-
    di       = @. li / ni   # grid step in x- and -y
    origin   = 0.0, 0.0     # origin coordinates
    xci, xvi = lazy_grid(di, li, ni; origin=origin) # nodes at the center and vertices of the cells
    dt       = Inf 
```

> Initialisation of the Rheology utilising [GeoParams.jl](https://github.com/JuliaGeodynamics/GeoParams.jl). The Rheology can be taylored to the specific problem with different creep laws and material parameters (see [GeoParams.jl](https://github.com/JuliaGeodynamics/GeoParams.jl)) or the miniapps in the [convection folder](miniapps/convection).

Reviewer: ~~Initialisation of the Rheology utilising GeoParams.jl~~ "Initialisation of the rheology with GeoParams.jl"

Note that `rheology` is no capitalized mid-sentence.

```julia
    # Physical properties using GeoParams ----------------
    τ_y     = 1.6           # yield stress. If do_DP=true, τ_y stand for the cohesion: c*cos(ϕ)
    ϕ       = 30            # friction angle
    C       = τ_y           # Cohesion
    η0      = 1.0           # viscosity
    G0      = 1.0           # elastic shear modulus
    Gi      = G0/(6.0-4.0)  # elastic shear modulus perturbation
    εbg     = 1.0           # background strain-rate
    η_reg   = 8e-3          # regularisation "viscosity"
    dt      = η0/G0/4.0     # assumes Maxwell time of 4
    el_bg   = ConstantElasticity(; G=G0, Kb=4)
    el_inc  = ConstantElasticity(; G=Gi, Kb=4)
    visc    = LinearViscous(; η=η0) 

>     pl      = DruckerPrager_regularised(;  # non-regularized plasticity

Reviewer: Cite one of Thibault's plascitiy papers (e.g. https://agupubs.onlinelibrary.wiley.com/doi/pdf/10.1029/2022GC010675)

Author: see L:175

Reviewer: Can you share the permalink of that line? I don't see anything

        C    = C,
        ϕ    = ϕ, 
        η_vp = η_reg,
        Ψ    = 0,
    ) 
    rheology = (

>         # Low density phase

Reviewer: ~~# Low density phase~~ # Matrix phase

        SetMaterialParams(;
            Phase             = 1,
            Density           = ConstantDensity(; ρ = 0.0),
            Gravity           = ConstantGravity(; g = 0.0),
            CompositeRheology = CompositeRheology((visc, el_bg, pl)),
            Elasticity        = el_bg,

        ),

>         # High density phase

Reviewer: ~~# High density phase~~ # Inclusion phase

        SetMaterialParams(;
            Density           = ConstantDensity(; ρ = 0.0),
            Gravity           = ConstantGravity(; g = 0.0),
            CompositeRheology = CompositeRheology((visc, el_inc, pl)),
            Elasticity        = el_inc,
        ),
    )
```
Initialisation of the Stokes arrays and the necesseary allocations. The rheology is computed with `compute_viscosity!` which is a function from `JustRelax.jl` and computes the viscosity according to the strain rate and the phase ratios.

```julia
    # Initialize phase ratios -------------------------------
    radius       = 0.1
    phase_ratios = PhaseRatio(ni, length(rheology))
    init_phases!(phase_ratios, xci, radius)

    # STOKES ---------------------------------------------
    # Allocate arrays needed for every Stokes problem
    stokes    = StokesArrays(ni, ViscoElastic)

>     pt_stokes = PTStokesCoeffs(li, di; ϵ=1e-6,  CFL = 0.75 / √2.1)

Reviewer: It would be good here state that the PT coefficients come from [Ludo's et al GMD paper](https://gmd.copernicus.org/articles/15/5757/2022/)

```
Pseudo-transient Stokes solver and visualisation of the results. The visualisation is done with [GLMakie.jl](https://github.com/MakieOrg/Makie.jl). 

```julia    
    # IO ------------------------------------------------
    # if it does not exist, make folder where figures are stored
    !isdir(figdir) && mkpath(figdir)
    # ----------------------------------------------------

    # Time loop
    t, it      = 0.0, 0
    tmax       = 3.5
    τII        = Float64[]
    sol        = Float64[]
    ttot       = Float64[]

    while t < tmax

        # Stokes solver ----------------
        solve!(
            stokes,
            pt_stokes,
            di,
            flow_bcs,
            ρg,
            η,
            η_vep,
            phase_ratios,
            rheology,
            args,
            dt,
            igg;
            verbose          = false,
            iterMax          = 500e3,
            nout             = 1e3,
            viscosity_cutoff = (-Inf, Inf)
        )
        @parallel (JustRelax.@idx ni) JustRelax.Stokes2D.tensor_invariant!(stokes.ε.II, @strain(stokes)...)
        push!(τII, maximum(stokes.τ.xx))

>         if !isinf(dt)

Reviewer: remove `if !isinf(dt)`, the lines within this block should always run.

* [GeoParams.jl](https://github.com/JuliaGeodynamics/GeoParams.jl)
* [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl)

The package serves several purposes:

  * It provides a collection of solvers to be used in quickly developing new applications
  * It provides some standardization so that application codes can

     - more easily handle local material properties through the use of [GeoParams.jl]((https://github.com/JuliaGeodynamics/GeoParams.jl))
     - more easily switch between a pseudo-transient solver and another solvers (e.g. an explicit thermal solvers)

  * It provides a natural repository for contributions of new solvers for use by the larger community 

We provide several miniapps, each designed to solve a well-specified benchmark problem, in order to provide

  - examples of usage in high-performance computing
  - basis on which to build more full-featured application codes
  - cases for reference and performance tests

## Installation

`JustRelax.jl` is not yet registered (we are in the process), however it can be installed from the repository

```julia
using Pkg; Pkg.add("https://github.com/PTsolvers/JustRelax.jl.git")
```

>  If you have navigated the terminal to the directory you cloned `JustRelax.jl`, you can test the package by running the following commands:

Reviewer: ```suggestion
After installation, you can test the package by running the following commands:
```

```julia

> using JustRelax

Reviewer: ```suggestion
using JustRelax
julia> ] 
  pkg> test JustRelax
```

julia> ] 
  pkg> activate .
  pkg> instantiate
  pkg> test JustRelax
```
The test will take a while, so grab a :coffee: or :tea:

## Example: shear band localisation (2D)

![ShearBand2D](miniapps/benchmarks/stokes2D/shear_band/movies/DP_nx2058_2D.gif)

> This example benchmark test displays how the package can be used to simulate shear bands. The example is based on the [ShearBands2D.jl](miniapps/benchmarks/stokes2D/shear_band/ShearBand2D.jl). 

Reviewer: ```suggestion
This example example displays how the package can be used to simulate shear band localisation. The example is based on the [ShearBands2D.jl](miniapps/benchmarks/stokes2D/shear_band/ShearBand2D.jl). 
```

```julia
using GeoParams, GLMakie, CellArrays
using JustRelax, JustRelax.DataIO

# setup ParallelStencil.jl environment

> model  = PS_Setup(:CUDA, Float64, 2)

Reviewer: ```suggestion
model  = PS_Setup(:Threads, Float64, 2)
```

![CI](https://github.com/PTSolvers/JustRelax.jl/actions/workflows/ci.yml/badge.svg)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10212423.svg)](https://doi.org/10.5281/zenodo.10212423)

:warning: This Package is still under active development. The benchmarks and miniapps are working and provide the user with an insight into the capabilities of the package. However, the API is still subject to change. :warning: 

> Need to solve a very large multi-physics problem on a GPU cluster? Just Relax!

Reviewer 2: ```suggestion
Need to solve a very large multi-physics problem on many GPUs in parallel? Just Relax!
```

> Pseudo-transient accelerated iterative solvers, ready for extreme-scale, multi-GPU computation.

Reviewer 2: Maybe this line is now obsolete?

> `JustRelax.jl` is a collection of pseudo-transient relaxation solvers

Reviewer 2: ```suggestion
`JustRelax.jl` is a collection of accelerated iterative pseudo-transient solvers
```

for multi-physics problems on regular, staggered, parallel grids,

> using MPI and multiple CPU or GPU backends. It's part of the [PTSolvers project](https://ptsolvers.github.io) and

Reviewer 2: ```suggestion
using MPI and multiple CPU or GPU backends. It's part of the [PTSolvers organisation](https://ptsolvers.github.io) and
```

> the [GPU4GEO project](https://www.pasc-ch.org/projects/2021-2024/gpu4geo/). Current publications, outreach and news can be found on the [GPU4GEO website](https://ptsolvers.github.io/GPU4GEO/).

Reviewer 2: ```suggestion
developed within the [GPU4GEO project](https://www.pasc-ch.org/projects/2021-2024/gpu4geo/). Current publications, outreach and news can be found on the [GPU4GEO website](https://ptsolvers.github.io/GPU4GEO/).
```

The package relies on other packages as building blocks and parallelisation tools:

* [ParallelStencil.jl](https://github.com/omlins/ParallelStencil.jl)
* [ImplicitGlobalGrid.jl](https://github.com/omlins/ImplicitGlobalGrid.jl)
* [GeoParams.jl](https://github.com/JuliaGeodynamics/GeoParams.jl)
* [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl)

The package serves several purposes:

  * It provides a collection of solvers to be used in quickly developing new applications
  * It provides some standardization so that application codes can

     - more easily handle local material propoerties through the use of [GeoParams.jl]((https://github.com/JuliaGeodynamics/GeoParams.jl))

>      - more easily switch between a psuedo-transient solver and another solver (e.g. an explicit thermal solvers)

Reviewer 2: ```suggestion
     - more easily switch between a pseudo-transient solver and other solvers (e.g. an explicit thermal solvers)
```

>   * It provides a place to describe performance benchmarks for the solver routines

Reviewer: I think we can even delete this sentence.

>   * It provides a natural location for contributions of new solvers for use by the larger community

Reviewer 2: ```suggestion
  * It provides a natural repository for contributions of new solvers for use by the larger community
```

> We include several miniapps, each designed to solve a well-specified benchmark problem, in order to provide

Reviewer 2: ```suggestion
We provide several miniapps, each designed to solve a well-specified benchmark problem, in order to provide
```

>   - examples of high-performance usage,

Reviewer 2: ```suggestion
  - examples of usage in high-performance computing,
```

  - bases on which to build more full-featured application codes
  - cases for reference and performance tests

## Installation

> `JustRelax.jl` is not yet registered (we are in the process), however it can be installed by cloning the repository. If you have navigated the terminal to the directory you cloned `JustRelax.jl`, you can test the package by running the following commands:

Reviewer 2: ```suggestion
`JustRelax.jl` is not yet registered (we are in the process), however it can be installed by cloning the repository. If you have navigated the terminal to the directory where you cloned `JustRelax.jl`, you can test the package by running the following commands:
```

```julia
using JustRelax
julia> ] 
  pkg> activate .
  pkg> instantiate
  pkg> test JustRelax
```

> The test will take a while, so grab a coffee or tea. 

Reviewer 2: Maybe worth also mentioning how to add JustRelax in Julia from Pkg?

## Usage

> For the advection of physical fields, this package relies on the Particles-In-Cell method implemented in [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl). Depending on where you want to run `JustRelax.jl` (CPU or GPU) you will need to specify the Backend. Run this command in the REPL and restart Julia:

Reviewer 2: ```suggestion
JustRelax can execute on different backends, currently CPUs, and Nvidia and AMD GPUs. Execute the following command in the REPL to select the backend, and restart Julia:
```

Reviewer 2: Following "For the advection of physical fields, this package relies on the Particles-In-Cell method implemented in [JustPIC.jl](https://github.com/JuliaGeodynamics/JustPIC.jl). " I would put somewhere else.

```julia
  set_backend("Threads_Float64_2D")       #running on the CPU
  set_backend("CUDA_Float64_2D")          #running on an NVIDIA GPU
  set_backend("AMDGPU_Float64_2D")        #running on an AMD GPU
```

> After you have run your desired line and restarted Julia, there should be a file called `LocalPreferences.toml` in the directory together with your `Project.toml` and `Manifest.toml`. This file contains the information about the backend you want to use. If you want to change the backend, you can run the command again. The default backend is `Threads_Float64_2D`.

Reviewer 2: ```suggestion
After restarting Julia, there should be a file called `LocalPreferences.toml` in the directory together with your `Project.toml` and `Manifest.toml`. This file contains the information about the backend to use. To change the backend further, simply run the command again. The backend defaults to `Threads_Float64_2D`.
```

> As stated above, the parallelisation is done using ParallelStencil.jl. Therefore, `JustRelax.jl` has as an environment already setup with which you can specify the dimension of your problem (2D or 3D) and the backend (CPU or GPU). The `PS_setup` variables are `PS_Setup(device, precission, dimensions)`. If you therefore want to run a 3D code, change the `dimensions` to 3 in the available commands below.

Reviewer 2: ```suggestion
JustRelax allows to setup a model environment `PS_Setup` (and  interplay with the underlying ParallelStencil package) to specify the dimension of the problem (2D or 3D) and the backend (CPU or GPU). The `PS_setup` functions takes `device`, `precision` and `dimensions` as argument:
```

Author: thanks, updated accordingly

```julia
  model = PS_Setup(:Threads, Float64, 2)  #running on the CPU in 2D
  environment!(model)

  model = PS_Setup(:CUDA, Float64, 2)     #running on an NVIDIA GPU in 2D
  environment!(model)

  model = PS_Setup(:AMDGPU, Float64, 2)   #running on an AMD GPU in 2D
  environment!(model)
```

## Shear Band 2D example

This example benchmark test displays how the package can be used to simulate shear bands. The example is based on the [ShearBands2D.jl](miniapps/benchmarks/stokes2D/shear_band/ShearBand2D.jl). 

![ShearBand2D](miniapps/benchmarks/stokes2D/shear_band/movies/DP_nx2058_2D.gif)

> The `CUDA.allowscalar(false)` disallows scalar indexing. This could be regarded as a safety measure, as the scalar indexing would run super slow on the CPU with CUDA and will actually crash with AMD. 

Reviewer 2: ```suggestion
The `CUDA.allowscalar(false)` disallows scalar indexing. This could be regarded as a safety measure, as scalar indexing is highly inefficient or may even error on GPUs. 
```

```julia

N      = 128
n      = N + 2
nx     = n - 2
ny     = n - 2
figdir = "ShearBands2D"
igg  = if !(JustRelax.MPI.Initialized())
    IGG(init_global_grid(nx, ny, 0; init_MPI = true)...)
else
    igg
end
main(igg; figdir = figdir, nx = nx, ny = ny);
```

## Miniapps

Currenlty there are 3 convection miniapps with particles and 3 corresponding miniapps without. The miniapps with particles are:

  * [Layered_convection2D.jl](miniapps/convection/Particles2D/Layered_convection2D.jl)
  * [Layered_convection3D.jl](miniapps/convection/Particles3D/Layered_convection3D.jl)
  * [WENO_convection2D.jl](miniapps/convection/WENO5/WENO_convection2D.jl)

The miniapps without particles are:
  * [GlobalConvection2D_Upwind.jl](miniapps/convection/GlobalConvection2D_Upwind.jl)
  * [GlobalConvection3D_Upwind.jl](miniapps/convection/GlobalConvection3D_Upwind.jl)
  * [GlobalConvection2D_WENO5.jl](miniapps/convection/GlobalConvection2D_WENO5.jl)

  ## Benchmarks

>   We got multiple benchmark tests for Stokes in 2D and 3D as well as thermal diffusion. The benchmarks are updated to be up to date to the current Solvers and version. The benchmarks can be found in the [Benchmarks](miniapps/benchmarks).

Reviewer 2: ```suggestion
Current (Stokes 2D-3D, thermal diffusion) and future benchmarks can be found in the [Benchmarks](miniapps/benchmarks).
```
